LogQL: Pattern Parser (#3837)
grafana/loki: docs/sources/logql/_index.md

If an extracted label key name already exists in the original log stream, the extracted label key will be suffixed with the `_extracted` keyword to make the distinction between the two labels. You can forcefully override the original label using a [label formatter expression](#labels-format-expression). However if an extracted key appears twice, only the latest label value will be kept.

We support currently support [json](#json), [logfmt](#logfmt), [pattern](#pattern), [regexp](#regexp) and [unpack](#unpack) parsers.

> It's easier to use the predefined parsers like `json` and `logfmt` when you can, otherwise `pattern` and `regexp` parser can be used when the log lines have unusual structure. However the `pattern` parser is easier and faster than the `regexp` one.

Reviewer: > However the `pattern` parser is easier and faster than the `regexp` one.

I think we might want to specify that it's easier and faster _to write_. It might also be nice to say that it's faster in terms of performance, too.

> It's easier to use the predefined parsers like `json` and `logfmt` when you can, otherwise `pattern` and `regexp` parser can be used when the log lines have unusual structure. However the `pattern` parser is easier and faster to writer, but also perform better than the `regexp` one.

Reviewer: ```suggestion
It's easier to use the predefined parsers like `json` and `logfmt` when you can, otherwise `pattern` and `regexp` parser can be used when the log lines have unusual structure. However the `pattern` parser is easier and faster to write, but also perform better than the `regexp` one.
```
:slightly_smiling_face:

"status" => "200"
```

##### pattern

> The pattern parser allow to explicitly extract fields from your log lines by defining a pattern expression which matched how the log line is.

Reviewer: ```suggestion
The pattern parser allow to explicitly extract fields from your log lines by defining a pattern expression which matches how the log line is structured.
```

> For example the following nginx log line:

Reviewer: Can we specify that this is whitespace-sensitive, and that whitespace is required to distinguish capture groups?

Author: it's actually literals that distinguish capture group.

"uri" => "/api/plugins/versioncheck"
"status" => "200"
"size" => "2"
"agent" => "Go-http-client/2.0"
```

A pattern expression is composed of captures and literals.

A capture is noted `<example>` where `example` is the name of the field that will be extracted.
Captures can be unnamed using `<_>`, in which case the capture skips matched content.

Captures are matched from the beginning or the previous set of literals, to the end or the next set of literals.
If a capture is not matched, the pattern parser will stop.

By default a pattern expression is anchored at the start, this means if the expression start with literals, the log line must also start with the same set of literals. You can use `<_>` at the beginning if you do not want to anchor your expression at the start.

For example, given the following log line:

```log
level=debug ts=2021-06-10T09:24:13.472094048Z caller=logging.go:66 traceID=0568b66ad2d9294c msg="POST /loki/api/v1/push (204) 16.652862ms"
```

If you want to match from `msg="`, you can use the following expression:

```pattern
<_> msg="<method> <path> (<status>) <latency>"
```

A pattern expression is invalid if:

- it does not contain any named capture.

> - it contains two consecutive captures.

Reviewer: ```suggestion
- it contains two consecutive captures not separated by whitespace.
```

"status" => "200"
```

##### pattern

The pattern parser allow to explicitly extract fields from your log lines by defining a pattern expression which matches how the log line is structured.

For example the following nginx log line:

```log
0.191.12.2 - - [10/Jun/2021:09:14:29 +0000] "GET /api/plugins/versioncheck HTTP/1.1" 200 2 "-" "Go-http-client/2.0" "13.76.247.102, 34.120.177.193" "TLSv1.2" "US" ""
```

Can be parsed with the expression:

`<ip> - - <_> "<method> <uri> <_>" <status> <size> <_> "<agent>" <_>`

and will extract the following fields

```kv
"ip" => "0.191.12.2"
"method" => "GET"
"uri" => "/api/plugins/versioncheck"
"status" => "200"
"size" => "2"
"agent" => "Go-http-client/2.0"
```

A pattern expression is composed of captures and literals.

A capture is noted `<example>` where `example` is the name of the field that will be extracted.
Captures can be unnamed using `<_>`, in which case the capture skips matched content.

> Captures are matched from the beginning or the previous set of literals, to the end or the next set of literals.

Author: yes it is ! will do

If a capture is not matched, the pattern parser will stop.

Literals can be any sequence of UTF-8 characters, including whitespace.

> By default a pattern expression is anchored at the start, this means if the expression start with literals, the log line must also start with the same set of literals. You can use `<_>` at the beginning if you do not want to anchor your expression at the start.

Reviewer 2: ```suggestion
By default, a pattern expression is anchored at the start of the log line. If the expression start with literals, then the log line must also start with the same set of literals. Use `<_>` at the beginning of the expression to anchor the expression at the start.
```

> For example, given the following log line:
> 
> ```log
> level=debug ts=2021-06-10T09:24:13.472094048Z caller=logging.go:66 traceID=0568b66ad2d9294c msg="POST /loki/api/v1/push (204) 16.652862ms"
> ```
> 
> If you want to match from `msg="`, you can use the following expression:
> 
> ```pattern
> <_> msg="<method> <path> (<status>) <latency>"
> ```

Reviewer 2: ```suggestion
Consider the log line

```log
level=debug ts=2021-06-10T09:24:13.472094048Z caller=logging.go:66 traceID=0568b66ad2d9294c msg="POST /loki/api/v1/push (204) 16.652862ms"
```

To match `msg="`, use the expression:

```pattern
<_> msg="<method> <path> (<status>) <latency>"
```
```

Author: hm the suggestion didn't work out

> A pattern expression is invalid if:
> 
> - it does not contain any named capture.
> - it contains two consecutive captures not separated by whitespace.

Reviewer 2: ```suggestion
A pattern expression is invalid if

- It does not contain any named capture.
- It contains two consecutive captures not separated by whitespace characters.
```
